desk/app/verifier.hoon

```diff
@@ -0,0 +1,1151 @@
+    ?-  -.req
+        %status
+      :^  %'POST'  (cat 3 base '/status')  heads
+      (make-body 'phoneNumber'^s+nr 'ship'^s+(scot %p who.req) ~)
+    ::
+        %verify
+      :^  %'POST'  (cat 3 base '/verify')  heads
+      (make-body 'phoneNumber'^s+nr ~)
+    ::
+        %submit
+      :^  %'PATCH'  (cat 3 base '/verify')  heads
+      (make-body 'phoneNumber'^s+nr 'otp'^s+otp.req ~)
+    ==
+  --
+::
+++  twitter
+  =>  [^twitter .]
+  |_  [=bowl:gall bearer=@t]
+  ++  req-post
+    |=  [handle=@t tweet=@t]
+    ^-  card
+    =;  =request:http
+      :+  %pass  /id/twitter/(scot %t handle)/post/(scot %t tweet)
+      [%arvo %i %request request %*(. *outbound-config:iris retries 0)]
+    =/  heads=header-list:http
+      ['authorization' (cat 3 'Bearer ' bearer)]~
+    =-  [%'GET' - heads ~]
+    ::TODO  query params could be tighter if we're only looking at plaintext body,
+    ::      don't need entities
+    %+  rap  3
+    :~  'https://api.x.com/2/tweets/'
+        tweet
+        '?user.fields=username,id'
+        '&tweet.fields=author_id,entities'
+        '&expansions=author_id'
+    ==
+  ::
+  ++  parse-post
+    |=  $:  [handle=@t nonce=@ux]
+            [hed=response-header:http bod=(unit mime-data:iris)]
+        ==
+    ^-  ?([%good sig=@ux] %rate-limited %unauthorized %not-found %protected %bad-handle %bad-tweet %bad-nonce %bad-sign %bad)
+    ?+  status-code.hed  %bad
+      %400  %bad
+      %401  %unauthorized
+      %404  %not-found
+      %429  %rate-limited
+    ::
+        %200
+      ?~  bod  %bad
+      ?~  jon=(de:json:html q.data.u.bod)  %bad
+      =,  dejs-soft:format
+      =/  [[text=@t uid=@t] usr=(map @t @t)]
+        =-  (fall - ['' ''] ~)
+        %.  u.jon
+        %-  ot
+        :~  'data'^(ot 'text'^so 'author_id'^so ~)
+          ::
+            =;  u  'includes'^(ot 'users'^u ~)
+            (cu ~(gas by *(map @t @t)) (ar (ot 'id'^so 'username'^so ~)))
+        ==
+      ?~  nom=(~(get by usr) uid)
+        =;  nauth  ?:(nauth %protected %bad)
+        ::TODO  do better
+        ?=(^ (find "not-authorized-for-resource" (trip q.data.u.bod)))
+      ::  we only ever operate on lowercased handles
+      ::
+      =.  u.nom  (crip (cass (trip u.nom)))
+      ?:  !=(u.nom handle)  %bad-handle
+      =/  pull
+        ::NOTE  the twitter api returns newlines as backslash-n, so newlines
+        ::      directly preceding the blob will be included in its result.
+        ::      however, for any valid jam, appending arbitrary bytes does not
+        ::      change the result of cueing it: cue simply does not read into the
+        ::      extraneous bytes. so we only need to worry about content directly
+        ::      _after_ the blob, and there newlines will properly demarcate the
+        ::      end of the blob (as will spaces and other non-siw:ab characters),
+        ::      because they start with a backslash.
+        |^  wrap
+        ++  wrap  %+  knee  *@  |.  ~+
```

Comment:
oof ouch owie

Reply:
I can't stop drinking the hoon hurting juice.

(fwiw, this is a very standard necessary pattern for parsers that include themselves. `+knee` lets us tell the type system the parser will produce "something like an atom", `|.` lets it punt on evaluating the parser, and the `~+` is added in for performance safeguarding.)

(Actually no clue why `+knee` doesn't just take a `mold` in place of the example `*`...)